sklearn/semi_supervised/self_training.py

```diff
@@ -0,0 +1,156 @@
+
+
+class SelfTraining(BaseEstimator, ClassifierMixin):
+
+    """Self-training classifier
+
+    Parameters
+    ----------
+    base_estimator : estimator object
+        An estimator object implementing `fit` and `predict_proba`.
+
+    threshold : float
+        Threshold above which predictions are added to the labeled dataset.
+        Should be in [0, 1].
+
+    max_iter : integer
+        Maximum number of iterations allowed. Should be greater than or equal
+        to 0.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn import datasets
+    >>> from sklearn.semi_supervised import SelfTraining
+    >>> from sklearn.svm import SVC
+    >>> svc = SVC(probability=True, gamma="auto")
+    >>> self_training_model = SelfTraining(svc)
+    >>> iris = datasets.load_iris()
+    >>> rng = np.random.RandomState(42)
+    >>> random_unlabeled_points = rng.rand(len(iris.target)) < 0.3
+    >>> labels = np.copy(iris.target)
+    >>> labels[random_unlabeled_points] = -1
+    >>> self_training_model.fit(iris.data, labels)
+    ... # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    SVC(...)
+
+    References
+    ----------
+    David Yarowsky. 1995. Unsupervised word sense disambiguation rivaling
+    supervised methods. In Proceedings of the 33rd annual meeting on
+    Association for Computational Linguistics (ACL '95). Association for
+    Computational Linguistics, Stroudsburg, PA, USA, 189-196. DOI:
+    https://doi.org/10.3115/981658.981684
+    """
+    def __init__(self,
+                 base_estimator,
+                 threshold=0.75,
+                 max_iter=100):
+        self.base_estimator = base_estimator
+        self.threshold = threshold
+        self.max_iter = max_iter
+
+    def fit(self, X, y):
+        """
+        Fits SelfTraining estimator to dataset
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            array representing the data
+        y : array-like, shape = (n_samples, 1)
+            array representing the labels
+
+        Returns
+        -------
+        self: returns an instance of self.
+        """
+        X, y = check_X_y(X, y)
+        _check_estimator(self.base_estimator)
+
+        if not 0 <= self.max_iter:
+            raise ValueError("max_iter must be >= 0")
+
+        # Data usable for supervised training
+        X_labeled = X[safe_mask(X, np.where(y != -1))][0]
+        y_labeled = y[safe_mask(y, np.where(y != -1))][0]
+
+        # Unlabeled data
+        X_unlabeled = X[safe_mask(X, np.where(y == -1))][0]
+        y_unlabeled = y[safe_mask(y, np.where(y == -1))][0]
+
+        iter = 0
+        while (len(X_labeled) < len(X) and iter < self.max_iter):
+            iter += 1
+            self.base_estimator.fit(X_labeled, y_labeled)
+
+            # Select prediction where confidence is above the threshold
+            pred = self.base_estimator.predict(X_unlabeled)
```

Comment:
I think we should just use the argmax of predict_proba rather than predicting twice.